### PR TITLE
Implement function unflatObject

### LINKED ISSUE #114: Implement function to de-flat objects by nested keys concatenation with 1st level keys
Test:
```js
metatests.test('Object: deflatFields', (test) => {
  const source = {
    nameFirst: 'Andrew',
    nameSecond: ' Johnson',
    old: true,
    avoid: [1, 2, 3],
    parentMother: 'Eva',
    parentFather: 'Adam',
  };
  const expected = {
    name: { first: 'Andrew', second: 'Johnson' },
    old: true,
    avoid: [1, 2, 3],
    parent: { mother: 'Eva', father: 'Adam' },
  };

  const result = metautil.flatObject(source);

  console.log(result);
  test.strictSame(result, expected);
  test.end();
});

metatests.test('Object: deflatFields with key names', (test) => {
  const source = {
    nameFirst: 'Andrew',
    nameSecond: ' Johnson',
    old: true,
    avoid: [1, 2, 3],
    parent: { mother: 'Eva', father: 'Adam' },
  };
  const expected = {
    name: { first: 'Andrew', second: 'Johnson' },
    old: true,
    avoid: [1, 2, 3],
    parentMother: 'Eva',
    parentFather: 'Adam',
  };

  const result = metautil.flatObject(source, 'name');

  console.log(result);
  test.strictSame(result, expected);
  test.end();
});


```

Create your implementation of `deflatObject` function in `lib/utilites.js`

### PR DESCRIPTION
**Pull request by Fr3ud**
Closes #114 

- [x] tests and linter show no problems (`npm t`)
- [x] tests are added/updated for bug fixes and new features
- [x] code is properly formatted (`npm run fmt`)
- [x] description of changes is added in CHANGELOG.md
- [x] update .d.ts typings

### CODE REVIEW

**Comment by tshemsedinov** [comment]
Give commit tiples max 50 chars long. And short (1 line) PR titles. @Fr3ud

**Comment by Fr3ud** [comment]
@tshemsedinov, done

**Review by georgolden — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 ## [Unreleased][unreleased]
 
+- Add function `deflatObject(source: object, fieldName: string): object`
```
Comment (tshemsedinov) [inline CHANGELOG.md:5]:
```suggestion
- Add function `unflatObject(source: object, fieldNames: Array<string>): object`
```

File: lib/utilities.js [hunk 2]
```diff
@@ -329,6 +329,41 @@ const jsonParse = (buffer) => {
   }
 };
 
+const deflatObject = (source, fieldName) => {
```
[the remark this answers is not in the document]
Comment (tshemsedinov) [reply lib/utilities.js:332]: `fieldNames` is required paramater

File: lib/utilities.js [hunk 3]
```diff
@@ -329,6 +329,41 @@ const jsonParse = (buffer) => {
...
+  const result = {};
+  for (const [key, value] of Object.entries(source)) {
+    if (typeof value === 'string') {
+      for (let i = 0; i < key.length; i++) {
+        if (inRange(key[i], 'A', 'Z')) {
```
Comment (tshemsedinov) [inline lib/utilities.js:337]: Do not iterate by chars, just `for (const prefix of fieldNames)` and use `startsWith`

File: lib/utilities.js [hunk 4]
```diff
@@ -329,6 +329,41 @@ const jsonParse = (buffer) => {
...
+          }
+          if (source[firstWord] !== undefined) {
+            throw new Error(`${firstWord} key already exist`);
+          }
+          result[firstWord] = { ...result[firstWord], [secondWord]: value };
+        }
+      }
+      continue;
+    }
+    if (fieldName && typeof value === 'object' && !Array.isArray(value)) {
```
Comment (tshemsedinov) [inline lib/utilities.js:352]: We do not support arrays

File: lib/utilities.js [hunk 5]
```diff
@@ -356,6 +358,47 @@ const jsonParse = (buffer) => {
   }
 };
 
+const getPrefix = (key, fieldNames) => {
+  for (const prefix of fieldNames) {
+    if (key.startsWith(prefix)) {
+      return prefix;
+    }
```
Comment (tshemsedinov) [inline lib/utilities.js:365]:
```suggestion
    if (key.startsWith(prefix)) return prefix;
```

File: lib/utilities.js [hunk 6]
```diff
@@ -356,6 +358,47 @@ const jsonParse = (buffer) => {
...
+  }
+  return null;
+};
+
+const unflatObject = (source, fieldNames) => {
+  const result = {};
+  for (const [key, value] of Object.entries(source)) {
+    const prefix = getPrefix(key, fieldNames);
```
Comment (tshemsedinov) [inline lib/utilities.js:373]: Use .find

File: lib/utilities.js [hunk 7]
```diff
@@ -356,6 +358,47 @@ const jsonParse = (buffer) => {
...
+    if (prefix) {
+      if (source[prefix] !== undefined) {
```
Comment (tshemsedinov) [inline lib/utilities.js:375]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty

File: lib/utilities.js [hunk 8]
```diff
@@ -356,6 +358,47 @@ const jsonParse = (buffer) => {
...
+        throw new Error(`Can not combine keys: key "${prefix}" already exists`);
+      }
+      const newKey = key.slice(prefix.length).toLowerCase();
+      if (result[prefix] === undefined) {
+        result[prefix] = {};
+      }
```
Comment (tshemsedinov) [inline lib/utilities.js:381]:
```suggestion
      result[prefix] = result[prefix] || {};
```

**Comment by tshemsedinov** [comment]
See https://github.com/metarhia/metautil/pull/115 to resolve conflicts with landed `isHashObject`

File: lib/utilities.js [hunk 9]
```diff
@@ -388,6 +388,47 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
 };
 
+const unflatObject = (sourceObject, fieldNames = []) => {
+  const result = {};
+
+  for (const [key, value] of Object.entries(sourceObject)) {
+    const prefix = fieldNames.find((prefix) => key.startsWith(prefix));
+
+    if (prefix) {
+      // eslint-disable-next-line no-prototype-builtins
```
Comment (tshemsedinov) [inline lib/utilities.js:398]:
```suggestion
```

File: lib/utilities.js [hunk 10]
```diff
@@ -388,6 +388,47 @@ const flatObject = (sourceObject, fieldNames = []) => {
   return target;
 };
 
+const unflatObject = (sourceObject, fieldNames = []) => {
+  const result = {};
+
+  for (const [key, value] of Object.entries(sourceObject)) {
+    const prefix = fieldNames.find((prefix) => key.startsWith(prefix));
```
Comment (tshemsedinov) [inline lib/utilities.js:395]:
```suggestion
    const prefix = fieldNames.find((name) => key.startsWith(name));
```

File: lib/utilities.js [hunk 11]
```diff
@@ -388,6 +388,47 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
+
+    if (prefix) {
+      // eslint-disable-next-line no-prototype-builtins
+      if (sourceObject.hasOwnProperty(prefix)) {
```
Comment (tshemsedinov) [inline lib/utilities.js:399]:
```suggestion
      if (Object.prototype.hasOwnProperty.call(sourceObject, prefix) {
```

File: lib/utilities.js [hunk 12]
```diff
@@ -388,6 +388,47 @@ const flatObject = (sourceObject, fieldNames = []) => {
   return target;
 };
 
+const unflatObject = (sourceObject, fieldNames = []) => {
+  const result = {};
+
+  for (const [key, value] of Object.entries(sourceObject)) {
+    const prefix = fieldNames.find((prefix) => key.startsWith(prefix));
+
```
Comment (tshemsedinov) [inline lib/utilities.js:396]:
```suggestion
```

File: lib/utilities.js [hunk 13]
```diff
@@ -388,6 +388,47 @@ const flatObject = (sourceObject, fieldNames = []) => {
   return target;
 };
 
+const unflatObject = (sourceObject, fieldNames = []) => {
+  const result = {};
+
```
Comment (tshemsedinov) [inline lib/utilities.js:393]:
```suggestion
```

File: lib/utilities.js [hunk 14]
```diff
@@ -388,6 +388,47 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
+    const prefix = fieldNames.find((prefix) => key.startsWith(prefix));
+
+    if (prefix) {
+      // eslint-disable-next-line no-prototype-builtins
+      if (sourceObject.hasOwnProperty(prefix)) {
+        throw new Error(`Can not combine keys: key "${prefix}" already exists`);
+      }
+
+      const newKey = key.slice(prefix.length).toLowerCase();
+
```
Comment (tshemsedinov) [inline lib/utilities.js:404]:
```suggestion
      const newKey = key.slice(prefix.length).toLowerCase();
```

File: lib/utilities.js [hunk 15]
```diff
@@ -388,6 +388,47 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
+      result[prefix] = result[prefix] || {};
+      result[prefix][newKey] = value;
+      continue;
+    }
+
```
Comment (tshemsedinov) [inline lib/utilities.js:409]:
```suggestion
```

File: lib/utilities.js [hunk 16]
```diff
@@ -388,6 +388,47 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
+    if (isHashObject(value)) {
+      for (const [nestedKey, nestedValue] of Object.entries(value)) {
+        const newKey = `${key}${toUpperCamel(nestedKey)}`;
+        // eslint-disable-next-line no-prototype-builtins
+        if (sourceObject.hasOwnProperty(newKey)) {
+          throw new Error(
+            `Can not combine keys: key "${newKey}" already exists`,
+          );
+        }
```
Comment (tshemsedinov) [inline lib/utilities.js:418]:
```suggestion
        if (Object.prototype.hasOwnProperty.call(sourceObject, newKey)) {
          throw new Error(
            `Can not combine keys: key "${newKey}" already exists`,
          );
        }
```

File: lib/utilities.js [hunk 17]
```diff
@@ -388,6 +388,47 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
+      }
+
+      continue;
+    }
+
+    result[key] = value;
+  }
+
+  return result;
+};
```
Comment (tshemsedinov) [inline lib/utilities.js:430]:
```suggestion
        result[newKey] = nestedValue;
      }
      continue;
    }
    result[key] = value;
  }
  return result;
};
```

File: lib/utilities.js [hunk 18]
```diff
@@ -388,6 +388,36 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
+  const result = {};
+  for (const [key, value] of Object.entries(sourceObject)) {
+    const prefix = fieldNames.find((name) => key.startsWith(name));
+    if (prefix) {
+      if (Object.prototype.hasOwnProperty.call(sourceObject, prefix) {
+        throw new Error(`Can not combine keys: key "${prefix}" already exists`);
+      }
+      const newKey = key.slice(prefix.length).toLowerCase();
+      result[prefix] = result[prefix] || {};
+      result[prefix][newKey] = value;
```
Comment (tshemsedinov) [inline lib/utilities.js:401]:
```suggestion
      const newKey = key.slice(prefix.length).toLowerCase();
      const section = result[prefix];
      if (section) section[newKey] = value;
      else result[prefix] = { [newKey]: value };
```

File: lib/utilities.js [hunk 19]
```diff
@@ -388,6 +388,36 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
+      continue;
+    }
+    if (isHashObject(value)) {
+      for (const [nestedKey, nestedValue] of Object.entries(value)) {
+        const newKey = `${key}${toUpperCamel(nestedKey)}`;
+        if (Object.prototype.hasOwnProperty.call(sourceObject, newKey)) {
+          throw new Error(
+            `Can not combine keys: key "${newKey}" already exists`,
+          );
```
Comment (tshemsedinov) [inline lib/utilities.js:410]:
```suggestion
          const error = `Can not combine keys: key "${newKey}" already exists`;
          throw new Error(error);
```

File: test/utilities.js [hunk 20]
```diff
@@ -514,3 +514,74 @@ metatests.test('Object: flatFields duplicate key', (test) => {
...
+    parentFather: 'Adam',
+  };
+
+  const result = metautil.unflatObject(source, fieldNames);
+
+  test.strictSame(result, expected);
+  test.end();
+});
+
+metatests.test('Object: unflatFields exception', (test) => {
```
Comment (tshemsedinov) [inline test/utilities.js:568]:
```suggestion
metatests.test('Object: unflatFields naming collision', (test) => {
```

File: test/utilities.js [hunk 21]
```diff
@@ -514,3 +514,74 @@ metatests.test('Object: flatFields duplicate key', (test) => {
...
+    parentMother: 'Eva',
+    parentFather: 'Adam',
+  };
+
+  const result = metautil.unflatObject(source, fieldNames);
+
+  test.strictSame(result, expected);
+  test.end();
+});
+
```
Comment (tshemsedinov) [inline test/utilities.js:567]:
```suggestion
```

File: lib/utilities.js [hunk 22]
```diff
@@ -388,6 +388,36 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
+      for (const [nestedKey, nestedValue] of Object.entries(value)) {
+        const newKey = `${key}${toUpperCamel(nestedKey)}`;
+        if (Object.prototype.hasOwnProperty.call(sourceObject, newKey)) {
+          const error = `Can not combine keys: key "${newKey}" already exists`;
+          throw new Error(error);
+        }
+        result[newKey] = nestedValue;
+      }
+      continue;
+    }
```
Comment (tshemsedinov) [inline lib/utilities.js:415]:
```suggestion
```

File: lib/utilities.js [hunk 23]
```diff
@@ -388,6 +388,25 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
 
+const unflatObject = (sourceObject, fieldNames = []) => {
+  const result = {};
+  for (const [key, value] of Object.entries(sourceObject)) {
+    const prefix = fieldNames.find((name) => key.startsWith(name));
+    if (prefix) {
+      if (Object.prototype.hasOwnProperty.call(sourceObject, prefix) {
+        throw new Error(`Can not combine keys: key "${prefix}" already exists`);
+      }
+      const newKey = key.slice(prefix.length).toLowerCase();
```
Comment (tshemsedinov) [inline lib/utilities.js:399]:
```suggestion
      const newKey = key.substring(0, prefix.length).toLowerCase();
```

File: lib/utilities.js [hunk 24]
```diff
@@ -388,6 +388,25 @@ const flatObject = (sourceObject, fieldNames = []) => {
   return target;
 };
 
+const unflatObject = (sourceObject, fieldNames = []) => {
+  const result = {};
+  for (const [key, value] of Object.entries(sourceObject)) {
+    const prefix = fieldNames.find((name) => key.startsWith(name));
+    if (prefix) {
+      if (Object.prototype.hasOwnProperty.call(sourceObject, prefix) {
```
Comment (tshemsedinov) [inline lib/utilities.js:396]:
```suggestion
      if (Object.prototype.hasOwnProperty.call(sourceObject, prefix)) {
```

File: lib/utilities.js [hunk 25]
```diff
@@ -388,6 +388,25 @@ const flatObject = (sourceObject, fieldNames = []) => {
   return target;
 };
 
+const unflatObject = (sourceObject, fieldNames = []) => {
```
Comment (tshemsedinov) [inline lib/utilities.js:391]:
```suggestion
const unflatObject = (sourceObject, fieldNames) => {
```

File: lib/utilities.js [hunk 26]
```diff
@@ -388,6 +388,25 @@ const flatObject = (sourceObject, fieldNames = []) => {
...
 
+const unflatObject = (sourceObject, fieldNames) => {
+  const result = {};
+  for (const [key, value] of Object.entries(sourceObject)) {
+    const prefix = fieldNames.find((name) => key.startsWith(name));
+    if (prefix) {
+      if (Object.prototype.hasOwnProperty.call(sourceObject, prefix)) {
+        throw new Error(`Can not combine keys: key "${prefix}" already exists`);
+      }
+      const newKey = key.substring(0, prefix.length).toLowerCase();
```
Comment (tshemsedinov) [inline lib/utilities.js:399]:
```suggestion
      const newKey = key.substring(prefix.length).toLowerCase();
```

**Pull request merged**